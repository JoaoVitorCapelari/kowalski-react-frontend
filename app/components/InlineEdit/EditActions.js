import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FaCheck from 'react-icons/lib/fa/check';
import TiTimes from 'react-icons/lib/ti/times';

const EditActionsWrapper = styled.div`
  display: flex;
  flex-grow: 0;
  position: absolute;
  bottom: -28px;
  right: 0px;
`;

const CheckChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 2px;
  margin-right: 1px;
`;

const DiscardChangesButton = styled.a`
  display: flex;
  color: inherit;
  margin-left: 1px;
  margin-right: 2px;
`;

function EditActions({ onConfirmChanges, onDiscardChanges }) {
  return (
    <EditActionsWrapper>
      <CheckChangesButton title="Confirm changes" onMouseDown={onConfirmChanges}>
        <FaCheck size={19} />
      </CheckChangesButton>
      <DiscardChangesButton title="Discard changes" onMouseDown={onDiscardChanges}>
        <TiTimes size={22} />
      </DiscardChangesButton>
    </EditActionsWrapper>
  );
}

EditActions.propTypes = {
  onConfirmChanges: PropTypes.func.isRequired,
  onDiscardChanges: PropTypes.func.isRequired,
};

export default EditActions;
