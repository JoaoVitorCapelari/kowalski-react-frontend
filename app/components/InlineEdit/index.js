/**
*
* InlineEdit
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import FaPencil from 'react-icons/lib/fa/pencil';
import { ClipLoader } from 'react-spinners';
import EditActions from './EditActions';

export const STATE_NORMAL = 'STATE_NORMAL';
export const STATE_HOVERED = 'STATE_HOVERED';
export const STATE_EDITING = 'STATE_EDITING';
export const STATE_SAVING = 'STATE_SAVING';

const Container = styled.div`
  display: flex;
  flex-grow: 1;
  outline: ${({ outlined }) => outlined ? '1px solid grey' : 'none'};
`;

const ContentContainer = styled.div`
  display: flex;
  flex-grow: 1;
  position: relative;
  width: 100%;
  ${({ saving }) => saving ? 'font-style: italic; color: grey;' : ''}
`;

const SideIconsContainer = styled.a`
  color: inherit;
  margin-right: 5px;
  flex-grow: 0;
`;

function InlineEdit({
  formatter, onCommit, onDiscard, state, children, value, onStateChange, tooltipText, notSetMsg,
}) {
  return (
    <Container
      outlined={state === STATE_HOVERED || state === STATE_EDITING}
      onMouseEnter={() => state === STATE_NORMAL && onStateChange(STATE_HOVERED)}
      onMouseLeave={() => state === STATE_HOVERED && onStateChange(STATE_NORMAL)}
      onClick={() => state !== STATE_SAVING && onStateChange(STATE_EDITING)}
      title={state === STATE_HOVERED || state === STATE_NORMAL ? tooltipText : undefined}
    >
      <ContentContainer saving={state === STATE_SAVING}>
        { state !== STATE_EDITING && (value ? formatter(value) : notSetMsg) }
        { state === STATE_EDITING && children() }
        { state === STATE_EDITING && <EditActions onConfirmChanges={onCommit} onDiscardChanges={onDiscard} /> }
      </ContentContainer>
      {
        state !== STATE_EDITING && state !== STATE_NORMAL &&
          <SideIconsContainer>
            { state === STATE_HOVERED ? <FaPencil size={18} /> : <ClipLoader size={18} /> }
          </SideIconsContainer>
      }
    </Container>
  );
}

export const statePropType = PropTypes.oneOf([STATE_EDITING, STATE_NORMAL, STATE_HOVERED, STATE_SAVING]);

InlineEdit.propTypes = {
  formatter: PropTypes.func.isRequired,
  onCommit: PropTypes.func.isRequired,
  onDiscard: PropTypes.func.isRequired,
  state: statePropType.isRequired,
  children: PropTypes.func.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.instanceOf(Date)]),
  onStateChange: PropTypes.func.isRequired,
  tooltipText: PropTypes.string,
  notSetMsg: PropTypes.string,
};

InlineEdit.defaultProps = {
  formatter: (x) => x,
};

export default InlineEdit;
