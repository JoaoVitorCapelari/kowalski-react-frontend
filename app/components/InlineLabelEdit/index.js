/**
*
* InlineTextEdit
*
*/

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import InlineEdit from '../InlineEdit';

const Input = styled.input`
  width: 100%;
  height: 100%;
`;

function InlineLabelEdit(props) {
  return (
    <InlineEdit {...props}>
      { () => (
        <Input
          type="text"
          value={props.value}
          onChange={(evt) => props.onChange(evt.target.value)}
          innerRef={(ref) => {
            if (ref) {
              ref.focus();
              if (ref.value) {
                ref.selectionStart = ref.value.length + 1; // eslint-disable-line no-param-reassign
                ref.selectionEnd = ref.value.length + 1; // eslint-disable-line no-param-reassign
              }
            }
          }}
          onBlur={props.onCommit}
        />
      ) }
    </InlineEdit>
  );
}

InlineLabelEdit.propTypes = {
  onCommit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
};

export default InlineLabelEdit;
