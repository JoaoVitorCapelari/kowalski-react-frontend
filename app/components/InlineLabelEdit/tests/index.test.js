import React from 'react';
import { mount } from 'enzyme';

import InlineLabelEdit from '../index';
import { STATE_EDITING } from '../../InlineEdit';

describe('<InlineLabelEdit />', () => {
  let onCommit;
  let onChange;
  let onDiscard;
  let value;
  let componentWrapper;

  const renderComponent = () => mount(
    <InlineLabelEdit
      state={STATE_EDITING}
      onStateChange={jest.fn()}
      {...{ onCommit, onChange, onDiscard, value }}
    />
  );

  beforeEach(() => {
    onCommit = jest.fn();
    onChange = jest.fn();
    onDiscard = jest.fn();
    value = 'my testing label initial value';
    componentWrapper = renderComponent();
  });

  it('renders the value', () => {
    expect(componentWrapper.html()).toMatchSnapshot();
    expect(componentWrapper.html()).toContain(value);
  });

  describe('when the value is changed in the input', () => {
    const newValue = 'my new testing label';

    beforeEach(() => {
      componentWrapper.find('input').simulate('change', { target: { value: newValue } });
    });

    it('callback onChange properly', () => {
      expect(onChange).toHaveBeenCalledTimes(1);
      expect(onChange).toHaveBeenLastCalledWith(newValue);
    });
  });

  // it('callbacks onStateChange with STATE_HOVERED on mouseEnter', () => {
  //   const componentWrapper = renderComponent();
  //   componentWrapper.simulate('mouseEnter');
  //   expect(onStateChanged).toHaveBeenCalledTimes(1);
  //   expect(onStateChanged).toHaveBeenLastCalledWith(STATE_HOVERED);
  // });
});
