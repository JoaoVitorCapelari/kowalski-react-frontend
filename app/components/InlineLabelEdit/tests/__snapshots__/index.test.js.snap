// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineLabelEdit /> renders the value 1`] = `"<div class=\\"InlineEdit__Container-iwjTUk eHkAhw\\"><div class=\\"InlineEdit__ContentContainer-kvxWLC cdZpQT\\"><input type=\\"text\\" class=\\"InlineLabelEdit__Input-izYWLY eRqDwh\\" value=\\"my testing label initial value\\"><div class=\\"EditActions__EditActionsWrapper-ftJlAD iDsSJs\\"><a class=\\"EditActions__CheckChangesButton-bTUFoa gXLEtU\\" title=\\"Confirm changes\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"19\\" width=\\"19\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m37.3 12.6q0 0.9-0.6 1.6l-19.2 19.1q-0.6 0.7-1.5 0.7t-1.6-0.7l-11.1-11.1q-0.6-0.6-0.6-1.5t0.6-1.5l3.1-3q0.6-0.7 1.5-0.7t1.5 0.7l6.6 6.5 14.6-14.6q0.6-0.6 1.5-0.6t1.5 0.6l3.1 3q0.6 0.6 0.6 1.5z\\"></path></g></svg></a><a class=\\"EditActions__DiscardChangesButton-jzKdhF jclCtP\\" title=\\"Discard changes\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"22\\" width=\\"22\\" viewBox=\\"0 0 40 40\\" style=\\"vertical-align: middle;\\"><g><path d=\\"m29 11c-1.3-1.3-3.4-1.3-4.7 0l-4.3 4.3-4.3-4.3c-1.3-1.3-3.4-1.3-4.7 0-1.3 1.3-1.3 3.4 0 4.7l4.3 4.3-4.3 4.3c-1.3 1.3-1.3 3.4 0 4.7 0.6 0.7 1.5 1 2.3 1s1.7-0.3 2.4-1l4.3-4.3 4.3 4.3c0.7 0.7 1.5 1 2.4 1s1.7-0.3 2.3-1c1.3-1.3 1.3-3.4 0-4.7l-4.3-4.3 4.3-4.3c1.3-1.3 1.3-3.4 0-4.7z\\"></path></g></svg></a></div></div></div>"`;
